lasagne/layers/recurrent.py

```diff
@@ -171,58 +181,72 @@ def get_output_for(self, input, mask=None, **kwargs):
         input = input.dimshuffle(1, 0, 2)
         seq_len, num_batch, _ = input.shape
 
-        # Because the input is given for all time steps, we can precompute
-        # the inputs to hidden before scanning. First we need to reshape
-        # from (seq_len, batch_size, num_inputs) to
-        # (seq_len*batch_size, num_inputs)
-        input = T.reshape(input,
-                          (seq_len*num_batch, -1))
-        input_dot_W = helper.get_output(
-            self.input_to_hidden, input, **kwargs)
-
-        # reshape to original (seq_len, batch_size, num_units)
-
-        input_dot_W = T.reshape(input_dot_W,
-                                (seq_len, num_batch, -1))
+        if self.precompute_input:
+            # Because the input is given for all time steps, we can precompute
+            # the inputs to hidden before scanning. First we need to reshape
+            # from (seq_len, batch_size, num_inputs) to
+            # (seq_len*batch_size, num_inputs)
+            input = T.reshape(input,
+                              (seq_len*num_batch, -1))
+            input = helper.get_output(
+                self.input_to_hidden, input, **kwargs)
+
+            # reshape to original (seq_len, batch_size, num_units)
+            input = T.reshape(input, (seq_len, num_batch, -1))
+
+        def clone_and_compute_output(network, network_input, new_params):
+            # Gets the output from the given network and calculate the output.
+            # The original parameters are replaced with the parameters given
+            # in new_params. Allows for use of the strict setting in scan.
+            original_hid_pre = helper.get_output(
+                network, network_input, **kwargs)
+            original_params = helper.get_all_params(self.hidden_to_hidden)
+            new_hid_pre = theano.clone(
```

Comment:
Is it actually more efficient to call this `theano.clone` on every step of the `scan` instead of just using `strict=False`?  I'm not really familiar with why `strict=True` helps.